hook-consumer/src/consumer.rs

```diff
@@ -0,0 +1,490 @@
+/// was unsuccessful but we can still attempt a retry), or failed (if the request was unsuccessful and no more retries
+/// may be attempted).
+///
+/// A webhook job is considered retryable after a failing request if:
+/// 1. The job has attempts remaining (i.e. hasn't reached `max_attempts`), and...
+/// 2. The status code indicates retrying at a later point could resolve the issue. This means: 429 and any 5XX.
+///
+/// # Arguments
+///
+/// * `webhook_job`: The webhook job to process as dequeued from `hook_common::pgqueue::PgQueue`.
+/// * `request_timeout`: A timeout for the HTTP request.
+async fn process_webhook_job(
+    webhook_job: PgTransactionJob<'_, WebhookJobParameters>,
+    request_timeout: std::time::Duration,
+) -> Result<(), WebhookConsumerError> {
+    match send_webhook(
+        &webhook_job.job.parameters.method,
+        &webhook_job.job.parameters.url,
+        &webhook_job.job.parameters.headers,
+        webhook_job.job.parameters.body.clone(),
+        request_timeout,
+    )
+    .await
+    {
+        Ok(_) => {
+            webhook_job
+                .complete()
+                .await
+                .map_err(|error| WebhookConsumerError::PgJobError(error.to_string()))?;
+            Ok(())
+        }
+        Err(WebhookConsumerError::RetryableWebhookError {
+            reason,
+            retry_after,
+        }) => match webhook_job.retry(reason.to_string(), retry_after).await {
+            Ok(_) => Ok(()),
+            Err(PgJobError::RetryInvalidError {
+                job: webhook_job,
+                error: fail_error,
+            }) => {
+                webhook_job
+                    .fail(fail_error.to_string())
+                    .await
+                    .map_err(|job_error| WebhookConsumerError::PgJobError(job_error.to_string()))?;
+                Ok(())
+            }
+            Err(job_error) => Err(WebhookConsumerError::PgJobError(job_error.to_string())),
+        },
+        Err(error) => {
+            webhook_job
+                .fail(error.to_string())
+                .await
+                .map_err(|job_error| WebhookConsumerError::PgJobError(job_error.to_string()))?;
+            Ok(())
+        }
+    }
+}
+
+/// Make an HTTP request to a webhook endpoint.
+///
+/// # Arguments
+///
+/// * `method`: The HTTP method to use in the HTTP request.
+/// * `url`: The URL we are targetting with our request. Parsing this URL fail.
+/// * `headers`: Key, value pairs of HTTP headers in a `std::collections::HashMap`. Can fail if headers are not valid.
+/// * `body`: The body of the request. Ownership is required.
+/// * `timeout`: A timeout for the HTTP request.
+async fn send_webhook(
+    method: &HttpMethod,
+    url: &str,
+    headers: &collections::HashMap<String, String>,
+    body: String,
+    timeout: std::time::Duration,
+) -> Result<reqwest::Response, WebhookConsumerError> {
+    let client = reqwest::Client::new();
```

Comment:
If I'm reading the internals right this is truly constructing a fresh client from nothing every time. It doesn't have to be now, but I think we'll really want to re-use a pool of some kind since we'll be hammering the same servers and could reuse connections.

It looks like `reqwest` is building a `ClientBuilder` (which has pool support) under the hood, so maybe we can just make one of those toplevel and reuse it?

Reply:
Oh, `ClientBuilder` even takes a `timeout` that'll apply to all of them.

Reply:
Definitely could re-use the client, and I like the idea of configuring it once for all. Not sure about how effective re-using connections would be, as it would depend on our workload and pool size, but happy to try it out.